[ONNX] Tool to verify exported model discrepancy between sets of inputs

BowenBao opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #78323

A graph is exported for each set of inputs. The exported graphs are then compared
to each other, and discrepancies are reported. This function first checks the jit
graph, and then the onnx graph.

Unless otherwise specified, the jit/ONNX graph is expected to be the same, regardless
of the inputs it used for exporting. A discrepancy would imply the graph exported is
not accurate when running with other set of inputs, which will typically results in
runtime error or output mismatches.

Fixes https://github.com/microsoft/onnx-converters-private/issues/25

BowenBao committed:
[ONNX] Tool to verify exported model discrepancy between sets of inputs

A graph is exported for each set of inputs. The exported graphs are then compared
to each other, and discrepancies are reported. This function first checks the jit
graph, and then the onnx graph.

Unless otherwise specified, the jit/ONNX graph is expected to be the same, regardless
of the inputs it used for exporting. A discrepancy would imply the graph exported is
not accurate when running with other set of inputs, which will typically results in
runtime error or output mismatches.

[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX] Tool to verify exported model discrepancy between sets of inputs"


A graph is exported for each set of inputs. The exported graphs are then compared
to each other, and discrepancies are reported. This function first checks the jit
graph, and then the onnx graph.

Unless otherwise specified, the jit/ONNX graph is expected to be the same, regardless
of the inputs it used for exporting. A discrepancy would imply the graph exported is
not accurate when running with other set of inputs, which will typically results in
runtime error or output mismatches.

[ghstack-poisoned]

justinchuby commented on test/onnx/test_verification.py:
```diff
@@ -0,0 +1,67 @@
+
+
+class TestVerification(unittest.TestCase):
+    def test_check_jit_model_diff(self):
```
I would include the expected behavior as part of the test method name. To make what a test is testing clear: `test_<method or function name>_<expected behavior>_[when_<condition>]`.

e.g. `test_method_x_raises_error_when_dims_is_not_a_sequence`

justinchuby replied:
Same for other test methods

justinchuby commented on torch/onnx/verification.py:
```diff
@@ -3,18 +3,22 @@
 import torch
-from torch import Tensor
+import torch._C._onnx as _C_onnx
+from torch import _C, Tensor
```
Use `torch.Tensor`

BowenBao replied:
No strong opinion. `import Tensor` appears to be a convention in PyTorch. Perhaps due to its usage mainly in type annotations, hence following the typings import style. I'm changing to `torch.Tensor` now. As long as we are consistent I'm ok.

justinchuby commented on torch/onnx/verification.py:
```diff
@@ -243,6 +247,238 @@ def compare_ort_pytorch_model_with_input(input_args, input_kwargs):
+    export_modules_as_functions: Union[bool, Set[Type[torch.nn.Module]]] = False
+
+    def __init__(self, **kwargs):
+        self.__dict__.update(kwargs)
```
We can use a dataclass for this. `__init__` should have a clear signature users can see

justinchuby commented on torch/onnx/verification.py:
```diff
@@ -243,6 +247,238 @@ def compare_ort_pytorch_model_with_input(input_args, input_kwargs):
+        """See function :func:`_graph_diff`."""
+        return self.diff_report()
+
+    def _indent(self, s):
```
`format_diff_lines` or somthing like that for clarity?

BowenBao replied:
it's adding indent, "\t" to start of each line. I think the current naming is clearer ...

justinchuby replied:
SGTM

justinchuby commented on torch/onnx/verification.py:
```diff
@@ -243,6 +247,238 @@ def compare_ort_pytorch_model_with_input(input_args, input_kwargs):
+        """See function :func:`_graph_diff`."""
+        return self.diff_report()
+
+    def _indent(self, s):
```
Also consider param naming:

```suggestion
    def _indent(self, lines: str):
```

justinchuby commented on torch/onnx/verification.py:
```diff
@@ -243,6 +247,238 @@ def compare_ort_pytorch_model_with_input(input_args, input_kwargs):
+    def _indent(self, s):
+        return "\n".join(["\t" + line for line in s.splitlines()])
+
+    def diff_report(self):
```
```suggestion
    def diff_report(self) -> str:
```

justinchuby commented on torch/onnx/verification.py:
```diff
@@ -243,6 +247,238 @@ def compare_ort_pytorch_model_with_input(input_args, input_kwargs):
+        )
+        graph_diff_report = "Graph diff:\n" + self._indent("".join(graph_diff)) + "\n"
+
+        for n_ref, n_check in zip(graph_a.nodes(), graph_b.nodes()):
```
Clarity: I would rename n_ref and n_check for more descriptive names

BowenBao replied:
changing to node_a/b to be consistent. (stack_a/b as well)

justinchuby commented on torch/onnx/verification.py:
```diff
@@ -243,6 +247,238 @@ def compare_ort_pytorch_model_with_input(input_args, input_kwargs):
+
+        for n_ref, n_check in zip(graph_a.nodes(), graph_b.nodes()):
+            if str(n_ref) != str(n_check):
+                graph_diff_report += "First diverging operator:\n"
```
`graph_diff_report` can be a list that is later joined using `"\n".join()`. This allows python to not construct a new string every time it is modified

justinchuby commented on torch/onnx/verification.py:
```diff
@@ -243,6 +247,238 @@ def compare_ort_pytorch_model_with_input(input_args, input_kwargs):
+
+                graph_diff_report += source_printout
+
+                break
```
Just a Q: does this mean we only show the first difference?

BowenBao replied:
yes. the rest of nodes likely to be all different after first discrepancy. Updated phrasing in the function docstring to mention discrepancy of first pair of nodes.

justinchuby commented on torch/onnx/verification.py:
```diff
@@ -243,6 +247,238 @@ def compare_ort_pytorch_model_with_input(input_args, input_kwargs):
+    model: Union[torch.nn.Module, torch.jit.ScriptModule],
+    test_inputs: Sequence[Tuple[Tuple[Any, ...], Mapping[str, Any]]],
+    export_options: ExportOptions,
+):
```
```suggestion
) -> str:
```

justinchuby commented on torch/onnx/verification.py:
```diff
@@ -243,6 +247,238 @@ def compare_ort_pytorch_model_with_input(input_args, input_kwargs):
+
+def _check_jit_model_diff(
+    model: Union[torch.nn.Module, torch.jit.ScriptModule],
+    test_inputs: Sequence[Tuple[Tuple[Any, ...], Mapping[str, Any]]],
```
This param may need some explanation on what is expected

BowenBao replied:
adjusted naming and docstring.

BowenBao replied:
adjusted naming and docstring.

justinchuby commented on torch/onnx/verification.py:
```diff
@@ -243,6 +247,238 @@ def compare_ort_pytorch_model_with_input(input_args, input_kwargs):
+    test_inputs: Sequence[Tuple[Tuple[Any, ...], Mapping[str, Any]]],
+    export_options: ExportOptions,
+):
+    if len(test_inputs) <= 1:
```
nit: < 2, for clarity. Because we need 2 or more

justinchuby reviewed (commented):
Can't wait to use it. Thanks for creating this! I read half of `verification`. Will continue later

justinchuby reviewed (changes requested):
*just to update feedback status

justinchuby commented:
I wonder how best we can expose the diff results. If we raise errors, we should probably provide `assert_xxx` functions like `numpy.testing` or `torch.testing`. We may also return a data structure with diff info _and_ provide a function that raises assertion errors.

BowenBao committed:
Update on "[ONNX] Tool to verify exported model discrepancy between sets of inputs"


A graph is exported for each set of inputs. The exported graphs are then compared
to each other, and discrepancies are reported. This function first checks the jit
graph, and then the onnx graph.

Unless otherwise specified, the jit/ONNX graph is expected to be the same, regardless
of the inputs it used for exporting. A discrepancy would imply the graph exported is
not accurate when running with other set of inputs, which will typically results in
runtime error or output mismatches.

[ghstack-poisoned]

justinchuby commented on torch/onnx/verification.py:
```diff
@@ -243,6 +247,238 @@ def compare_ort_pytorch_model_with_input(input_args, input_kwargs):
+    if opset_version is None:
+        opset_version = _constants.onnx_default_opset
+    symbolic_helper._set_opset_version(opset_version)
+    symbolic_helper._set_onnx_shape_inference(True)
```
Same for this var

justinchuby commented on torch/onnx/utils.py:
```diff
@@ -564,6 +564,7 @@ def flatten(x):
 
 
 def _create_jit_graph(model, args):
+    model = _pre_trace_quant_model(model, args)
```
Just for my learning: Why is this line placed on the top?

justinchuby commented on torch/onnx/verification.py:
```diff
@@ -243,6 +247,238 @@ def compare_ort_pytorch_model_with_input(input_args, input_kwargs):
+    dynamic_axes: Optional[Mapping[str, Union[Mapping[int, str], Sequence[int]]]] = None
+    keep_initializers_as_inputs: Optional[bool] = None
+    custom_opsets: Optional[Mapping[str, int]] = None
+    export_modules_as_functions: Union[bool, Set[Type[torch.nn.Module]]] = False
```
These are class vars, are they accessible with `self.`?

BowenBao commented on torch/_C/__init__.pyi.in:
```diff
@@ -487,6 +489,7 @@ class Graph:
     def insertPoint(self) -> Node: ...
     def insertGraph(self, callee: Graph, inputs: List[Value]) -> List[Value]: ...
     def makeMultiOutputIntoTuple(self) -> None: ...
+    def nodes(self) -> List[Node]: ...
```
A lot more. It appears they are added in a on need basis.

justinchuby replied:
I created #78757

justinchuby replied:
Looks like nodes return an iterator?

BowenBao replied:
good catch!

BowenBao commented on test/onnx/test_verification.py:
```diff
@@ -0,0 +1,67 @@
+            r"First diverging operator:(.|\n)*"
+            r"prim::Constant(.|\n)*"
+            r"Reference source location:(.|\n)*"
+            r"Check source location:",
```
`_GraphDiff` is what I'm thinking to serve that purpose. I'm thinking to leave these options open for the future.

BowenBao commented on torch/onnx/verification.py:
```diff
@@ -243,6 +247,238 @@ def compare_ort_pytorch_model_with_input(input_args, input_kwargs):
+        self.graph_b = graph_b
+
+    def __str__(self):
+        """See function :func:`_graph_diff`."""
```
good catch, name changed

BowenBao commented on torch/onnx/verification.py:
```diff
@@ -243,6 +247,238 @@ def compare_ort_pytorch_model_with_input(input_args, input_kwargs):
+class ExportOptions:
+    """Arguments used by :func:`torch.onnx.export`.
+
+    TODO: Adopt this in `torch.onnx.export` api to replace keyword arguments.
```
Trying 1. with `_experimental.py`. Could be a good place to hold other future attempts, as long as it doesn't exhibit a tendency of becoming a place hording dead code ...

BowenBao committed:
Update on "[ONNX] Tool to verify exported model discrepancy between sets of inputs"


A graph is exported for each set of inputs. The exported graphs are then compared
to each other, and discrepancies are reported. This function first checks the jit
graph, and then the onnx graph.

Unless otherwise specified, the jit/ONNX graph is expected to be the same, regardless
of the inputs it used for exporting. A discrepancy would imply the graph exported is
not accurate when running with other set of inputs, which will typically results in
runtime error or output mismatches.

[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX] Tool to verify exported model discrepancy between sets of inputs"


A graph is exported for each set of inputs. The exported graphs are then compared
to each other, and discrepancies are reported. This function first checks the jit
graph, and then the onnx graph.

Unless otherwise specified, the jit/ONNX graph is expected to be the same, regardless
of the inputs it used for exporting. A discrepancy would imply the graph exported is
not accurate when running with other set of inputs, which will typically results in
runtime error or output mismatches.

[ghstack-poisoned]

justinchuby commented on torch/onnx/_experimental.py:
```diff
@@ -0,0 +1,28 @@
+
+
+@dataclass
+class ExportOptions:
```
Use `@dataclasses.dataclass`

BowenBao commented on torch/onnx/verification.py:
```diff
@@ -243,6 +247,238 @@ def compare_ort_pytorch_model_with_input(input_args, input_kwargs):
+    opset_version = export_options.opset_version
+    if opset_version is None:
+        opset_version = _constants.onnx_default_opset
+    symbolic_helper._set_opset_version(opset_version)
```
hmm should I use `_globals.GLOBALS` or `GLOBALS`?

justinchuby replied:
I don't have a strong opinion. I used `GLOBALS` in other places when the refactoring was done for our global vars. Personally I think GLOBALS is unique enough in the module that we can make an exception to the import naming rules. Except for being redundant, I don't think there's anything wrong with `_globals.GLOBALS`

BowenBao replied:
thx for explanation. SGTM, sticking to `GLOBALS` for consistency.

BowenBao committed:
Update on "[ONNX] Tool to verify exported model discrepancy between sets of inputs"


A graph is exported for each set of inputs. The exported graphs are then compared
to each other, and discrepancies are reported. This function first checks the jit
graph, and then the onnx graph.

Unless otherwise specified, the jit/ONNX graph is expected to be the same, regardless
of the inputs it used for exporting. A discrepancy would imply the graph exported is
not accurate when running with other set of inputs, which will typically results in
runtime error or output mismatches.

Fixes https://github.com/microsoft/onnx-converters-private/issues/25

[ghstack-poisoned]

BowenBao commented on test/onnx/test_verification.py:
```diff
@@ -0,0 +1,67 @@
+                return x + y.data
+
+        test_inputs = [
+            ((torch.randn(2, 3), torch.randn(2, 3)), {}),
```
Added `setUp()` to fix the seed.

justinchuby replied:
We can use pytorch's test class for seed setting. with `common_utils.TestCase`

justinchuby commented on torch/onnx/verification.py:
```diff
@@ -243,6 +246,219 @@ def compare_ort_pytorch_model_with_input(input_args, input_kwargs):
+
+    A graph is exported for each set of inputs. The exported graphs are then compared
+    to each other, and discrepancies of first pair of nodes are reported. This function
+    first checks the jit graph, and then the onnx graph.
```
I would mention that the function returns the jit graph if it finds any discrepancy, without bothering to look at the onnx graph (but phrased better)

BowenBao commented on test/onnx/test_verification.py:
```diff
@@ -0,0 +1,100 @@
+            r"First diverging operator:(.|\n)*"
+            r"onnx::Constant(.|\n)*"
+            r"Former source location:(.|\n)*"
+            r"Latter source location:",
```
A few reasons why I didn't use exact strings.
* Influenced by upstream graph printing changes that are unrelated.
* Stack trace env dependent.
* too long :).

justinchuby replied:
Sounds good!

BowenBao commented on torch/onnx/verification.py:
```diff
@@ -243,6 +246,219 @@ def compare_ort_pytorch_model_with_input(input_args, input_kwargs):
+    if opset_version is None:
+        opset_version = _constants.onnx_default_opset
+
+    export_modules_as_functions = utils._setup_trace_module_map(
```
For local function export, additional module information needs to be tracked. This tracer feature is by default turned off, and enabling it requires extra setup, which is what the function does.

justinchuby replied:
I see! Would it be helpful adding the explanation inline? It doesn’t seem directly apparent

BowenBao replied:
It can be added in the update for utils.py.

BowenBao commented on torch/onnx/verification.py:
```diff
@@ -243,6 +246,219 @@ def compare_ort_pytorch_model_with_input(input_args, input_kwargs):
+
+        if dynamic_axes is None:
+            dynamic_axes = {}
+        utils._validate_dynamic_axes(dynamic_axes, model, input_names, output_names)
```
I think it makes sense. The code is mimicking what `_export` is doing, so I'd avoid changing it now. I'd want to refactor(distill) the exporter context setup part to avoid this duplication.

justinchuby replied:
Sounds good

BowenBao commented on torch/onnx/verification.py:
```diff
@@ -243,6 +246,219 @@ def compare_ort_pytorch_model_with_input(input_args, input_kwargs):
+            )
+
+            if ref_onnx_graph is None:
+                ref_onnx_graph = onnx_graph
```
Agree. I'm changing it with a Callable argument. I think this change will be eventually needed, so doing it now.

BowenBao committed:
Update on "[ONNX] Tool to verify exported model discrepancy between sets of inputs"


A graph is exported for each set of inputs. The exported graphs are then compared
to each other, and discrepancies are reported. This function first checks the jit
graph, and then the onnx graph.

Unless otherwise specified, the jit/ONNX graph is expected to be the same, regardless
of the inputs it used for exporting. A discrepancy would imply the graph exported is
not accurate when running with other set of inputs, which will typically results in
runtime error or output mismatches.

Fixes https://github.com/microsoft/onnx-converters-private/issues/25

[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX] Tool to verify exported model discrepancy between sets of inputs"


A graph is exported for each set of inputs. The exported graphs are then compared
to each other, and discrepancies are reported. This function first checks the jit
graph, and then the onnx graph.

Unless otherwise specified, the jit/ONNX graph is expected to be the same, regardless
of the inputs it used for exporting. A discrepancy would imply the graph exported is
not accurate when running with other set of inputs, which will typically results in
runtime error or output mismatches.

Fixes https://github.com/microsoft/onnx-converters-private/issues/25

[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX] Tool to verify exported model discrepancy between sets of inputs"


A graph is exported for each set of inputs. The exported graphs are then compared
to each other, and discrepancies are reported. This function first checks the jit
graph, and then the onnx graph.

Unless otherwise specified, the jit/ONNX graph is expected to be the same, regardless
of the inputs it used for exporting. A discrepancy would imply the graph exported is
not accurate when running with other set of inputs, which will typically results in
runtime error or output mismatches.

Fixes https://github.com/microsoft/onnx-converters-private/issues/25

[ghstack-poisoned]

BowenBao commented:
@pytorchbot merge -g

BowenBao commented:
cc @garymm  for merge approval

garymm reviewed (approved):
@pytorchbot merge -g

justinchuby commented:
@pytorchbot merge -g

justinchuby commented on torch/onnx/verification.py:
```diff
@@ -243,6 +247,255 @@ def compare_ort_pytorch_model_with_input(input_args, input_kwargs):
+        """See function :func:`diff_report`."""
+        return self.diff_report()
+
+    def _indent(self, lines: str) -> str:
```
Can we use https://docs.python.org/3/library/textwrap.html for this?
